Write Dockerfile

tkkuehn opened the pull request:
This PR adds a dockerfile to the project that can be used to build a docker (and/or singularity container) with afids-auto-train and/or afids-auto-apply.

Note: In testing, I've found a singularity image built from this container to run `afids-auto-apply` really slowly (specifically the `c4d -rf-apply` call). I think it still works, but is worth investigating further.

Preemptive comment: It would be good to add a GitHub action to automatically build the container and push it to Dockerhub (something like [this](https://github.com/khanlab/tar2bids/blob/master/.github/workflows/push_container.yml)). I can add that if we want it for this PR. We may, however, want to discuss making an `afids` Dockerhub account or something along those lines, though.

Commit:
Create first draft dockerfile

Commit:
Condense RUN commands into one

Commit:
remove subject incl/excl files

Commit:
Strip participant selection stuff

kaitj commented on Dockerfile:
```diff
@@ -0,0 +1,55 @@
+        unzip=6.0-26 \
+        zlib1g-dev=1:1.2.11.dfsg-2+deb11u1 \
+    && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/* \
+    && pip install --no-cache-dir \
```
Would it be worth using something like poetry to install these dependencies or have it generate a `requirements.txt` file? The dependencies being installed here are different than what is currently being specified in the `pyproject.toml` and it would be good to keep those synced.

tkkuehn replied:
That is probably worth doing, I didn't think of it.

kaitj replied:
Not sure which would be easier / preferable:

1. Install poetry in docker and have it install that way.
2. Install poetry in the actions workflow to be created, have it create a requirements.txt file that gets copied over when creating the docker container and installing via pip.

tkkuehn replied:
1 seems preferable to me

kaitj reviewed (commented):
Looks good to me and thanks for adding this! A couple of minor comments:

* Keeping the python dependencies in sync (see in-line comment in `Dockerfile`)
* Similar note to above, Snakebids is on 0.4.1 in pyproject vs 0.6.0 in Dockerfile
* Is there an advantage to condensing the RUN into a single command vs splitting it up to be system dependencies, python dependencies, and external tools?

With regards to the it being slow to run, I already mentioned on Slack, but I think it may be `snakebids==0.6.0` related, but would need to dive deeper there. And yes, I think it would be worth having an action for creating and pushing the container - is there an afids dockerhub?

tkkuehn commented:
>     Is there an advantage to condensing the RUN into a single command vs splitting it up to be system dependencies, python dependencies, and external tools?

The idea here is to keep the layer count low (see [best practices doc](https://docs.docker.com/develop/dev-best-practices/#how-to-keep-your-images-small)), but on review it seems like smart use of multi-stage builds would be better and more readable -- I'll take another look.

> Is there an afids dockerhub?

Not that I'm aware of but possibly -- @greydongilmore do you know?

Commit:
update dependencies

- brings snakebids to 0.6.1 to resolve cookiecutter vulnerability
- removes redundant snakemake dependency (snakebids already installs
this)
- even though numpy and pandas is redundant in snakebids (due to
pybids), left in as it would still be good to be able to specify
specific versions for libraries that are directly used

Commit:
update poe tasks; quality dependencies

- updates black to v22.1 as well as black dependencies

Commit:
rm pipeline_description.json

This file was at the moment redundant with pyproject.toml, while also
not being duplicated in afids-auto-apply train. If we are to retain this
file, we should:

(1) Have it mirror the information from the pyproject.toml
(2) Either have it in a central location for both workflows, or have it
duplicated across the train and apply workflows.

Commit:
Rm unnecessary specs.json file

Commit:
quality fixes to .py files

Commit:
rm poetry.lock and pre-commit from .gitignore

Commit:
reset config files

- makes --model-dir a required argument for apply

Commit:
Use multi-stage build to install requirements

tkkuehn committed:
Merge pull request #10 from afids/cleanup

Repo cleanup

Commit:
Use poetry to install project requirements

kaitj reviewed (approved):
Looks good to me. One more question regarding the Dockerfile with respect to the multiple `ENTRYPOINT`s. Was wondering how that works to choose between `train` and `apply`? My understanding was that it sets the default command when you do something like `docker run` (and presumably `singularity run`)

tkkuehn commented:
Ah good point, should have documented that somewhere. What you can do with this is build two different containers buy invoking `docker build` differently:

- afids-auto-train: `sudo docker build . --target train -t afids-auto-train`
- afids-auto-apply: `sudo docker build . --target train -t afids-auto-apply`

Then you can push the build containers by their tags.

kaitj commented:
In any case, this looks good to me. One feature that would be nice to have down the line (and probably would replace the current entrypoints), is a command that can select either the `train` or `apply` workflows as needed. I think there might be some examples of that in an existing workflow - would have to take a look.
